### PR TITLE
tap: implement HTTP streaming tapping

### PR DESCRIPTION
**Pull request by mattklein123**
Users can now choose between buffered tapping (simpler) and
streaming tapping (more flexible but harder to work with).

Streaming tapping for the transport socket will be added in a
follow-up.

Signed-off-by: Matt Klein <mklein@lyft.com>

*Risk Level*: Medium for users of tap, no risk otherwise.
*Testing*: New UTs and integration tests.
*Docs Changes*: Added
*Release Notes*: N/A

### CODE REVIEW

File: source/extensions/common/tap/tap_matcher.h [hunk 1]
```diff
@@ -24,20 +24,26 @@ using MatcherPtr = std::unique_ptr<Matcher>;
...
  */
 class Matcher {
 public:
+  struct MatchStatus {
+    bool operator==(const MatchStatus& rhs) const {
+      return matches_ == rhs.matches_ && might_change_status_ == rhs.might_change_status_;
+    }
+
+    bool matches_{false};            // Does the matcher currently match?
+    bool might_change_status_{true}; // Is it possible for matches_ to change in subsequent updates?
```
Comment (htuch) [inline source/extensions/common/tap/tap_matcher.h]: Can you split this to a separate PR? This looks like it could be cleanly done standalone. Thanks.

File: docs/root/configuration/http_filters/tap_filter.rst [hunk 2]
```diff
@@ -165,11 +167,64 @@ the request/response sequence, the filter will match incrementally as the reques
...
+<envoy_api_field_service.tap.v2alpha.OutputConfig.streaming>` setting. If this setting is false
+(the default), Envoy will emit :ref:`fully buffered traces
+<envoy_api_msg_data.tap.v2alpha.HttpBufferedTrace>`. Users are likely to find this format easier
+to interact with for simple cases.
+
+In cases where fully buffered traces are not practical (e.g., very large request and responses,
+long lived streaming APIs, etc.), the streaming setting can be set to true, and Envoy will emit
+multiple :ref:`streamed trace segments <envoy_api_msg_data.tap.v2alpha.HttpStreamedTraceSegment>` for
+each tap. In this case, it is required that post-processing is performed to stitch all of the trace
+segments back together into a usable form. Also note that binary protobuf is not a self-delimiting
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply docs/root/configuration/http_filters/tap_filter.rst]: Correct, the gRPC streaming API is effectively going to ignore the file format option.

File: source/extensions/common/tap/tap_matcher.h (hunk 1 above)
Comment (mattklein123) [reply source/extensions/common/tap/tap_matcher.h]: OK

File: api/envoy/data/tap/v2alpha/http.proto [hunk 3]
```diff
@@ -31,3 +31,35 @@ message HttpBufferedTrace {
   // Response message.
   Message response = 2;
 }
+
+// A streamed trace segment. Multiple segments make up a full trace.
+message HttpStreamedTraceSegment {
+  // Wrapper for a set of headers.
+  message HeaderMap {
```
[the remark this answers is not in the document]
Comment (lizan) [reply api/envoy/data/tap/v2alpha/http.proto]: Yeah perhaps `api/envoy/api/v2/core/http.proto`, but it can be moved later. Whole of this might be useful for #5209.

**Comment by mattklein123** [comment]
@htuch @lizan merged and updated

File: api/envoy/service/tap/v2alpha/common.proto [hunk 4]
```diff
@@ -90,12 +90,20 @@ message OutputConfig {
...
   // default is 1KiB.
   google.protobuf.UInt32Value max_buffered_tx_bytes = 3;
+
+  // Indicates whether taps produce a single buffered message per tap, or multiple streamed
+  // messages per tap in the emitted :ref:`TraceWrapper
+  // <envoy_api_msg_data.tap.v2alpha.TraceWrapper>` messages. Note that streamed tapping does not
+  // mean that no buffering takes place. Buffering may be required if data is processed before a
+  // match can be determined. See the HTTP tap filter :ref:`streaming
+  // <config_http_filters_tap_streaming>` documentation for more information.
+  bool streaming = 4;
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/service/tap/v2alpha/common.proto]: We should have some better checks here. I already have TODOs for some existing checks, but I will add some more and do this in a follow up.
